### DISCUSSION TITLE
password decryption doesn't fail 0.4% of the time while given the wrong password (depends on salt)

### DISCUSSION THREAD

**Discussion by lacostej** [openssl/openssl#23198] [Q&A] [answered]
In the @fastlane/fastlane (ruby) project, we use openssl password based encryption to encrypt the files we fetch from the Apple Developer portal. Under the hood, the ruby code generates an IV and Key using `EVP_BytesToKey` and generates a random salt.

If the wrong password is used, the code is supposed to fail with the `bad decrypt` error.

We found out that in [about 0.4% of the time](https://github.com/fastlane/fastlane/issues/21663), the decryption doesn't detect that the wrong password was entered. This depends on the salt that was generated.

We have a simple [ruby code](https://github.com/fastlane/fastlane/issues/21663#issuecomment-1875392279) demonstrating the issue.

We've [converted it to C and it appears to fail in the same way](https://github.com/fastlane/fastlane/issues/21663#issuecomment-1876700781). I haven't written C in 20 years so I hope I got this right :)

Is there an issue in openssl?

The ruby code was tested on various systems, from Mac to Linux.

The C code was tested on my Mac M1 Pro against openssl 3.2.0

**Comment by tom-cosgrove-arm** [comment]
Can you supply an encrypted file together with the correct password and a wrong password that works on that file?

**Comment by mattcaswell** [comment, marked as answer]
Firstly: don't use `EVP_BytesToKey()`...we should deprecate it (actually I thought it already was deprecated and was surprised to learn that it isn't). There are much better standards based KDF functions to use - this one is just an OpenSSL specific (non standardised) legacy function. I'd recommend PBKDF2 instead:

https://www.openssl.org/docs/man3.2/man7/EVP_KDF-PBKDF2.html

If you do use `EVP_BytesToKey` then don't use an iteration count of 1. Hopefully, you're just doing that for the sake of the test and to speed things up.

Anyway none of that has anything to do with why you see this behaviour.

`EVP_BytesToKey()` generates a key/iv from the given salt/password (whether or not that password is correct or not). You then use that key to decrypt the data. The encryption algorithm you use is AES-256-CBC which has no integrity checking built in. It will just decrypt the data with the given key and give you the output. It might be correct data, it might be garbage (if the key is wrong).

The plaintext data is always padded before encryption to make it a multiple of the block length (16 bytes). If the input data is 15 bytes long, then it will pad with the bytes `0x01`. If it is 14 bytes it will pad with `0x02 0x02`, and so on. This padding is stripped off during decryption.

The decryption functions will only fail if the padding is invalid after decryption. For example if the last byte `>0x10` then this can never be valid padding. If the last byte is `0x02`, then the preceding byte must also be `0x02` - otherwise this is also invalid, etc.

If the last byte is `0x01` then this is always valid padding and the data will successfully decrypt. There is a 1 in 256 chance of the garbage data that is decrypted in the event of an incorrect key ending in `0x01`, i.e. approximately 0.4%.

So, this is not an OpenSSL bug. It is expected behaviour. The AES-256-CBC algorithm does not give you any guarantees of the decryption failing in the event of a wrong key. If you want that behaviour then you need to implement some integrity checking (for example you could use an AEAD based cipher such as AES-256-GCM).

**Reply by lacostej** [reply]
Here's a [potential implementation in ruby](https://gist.github.com/lacostej/cc33868553e0d57f27681dc887865caa) that uses AES-256-GSM. This time it uses a high number of iterations and [derive](https://gist.github.com/lacostej/cc33868553e0d57f27681dc887865caa#file-fastlane_openssl_migration_test-rb-L62-L70) the `key`, `iv` and `auth_data` from the `password` and `salt`.

Do you see any drawback in the `MatchDataEncryption` proposed implementation? E.g. in particular in [storing ](https://gist.github.com/lacostej/cc33868553e0d57f27681dc887865caa#file-fastlane_openssl_migration_test-rb-L79-L81) and [recovering](https://gist.github.com/lacostej/cc33868553e0d57f27681dc887865caa#file-fastlane_openssl_migration_test-rb-L91-L97) the `auth_tag` alongside the `salt` in the stored data string, just before the encrypted data.

Thanks.

**Reply by mattcaswell** [reply]
I don't know Ruby, or the OpenSSL Ruby bindings - but at the level that I can read it, it looks reasonable on an initial scan through. Interestingly I note that the file prefix you are using ("Salted__") matches exactly the prefix that the "openssl enc" command line tool uses for its password encrypted files. If file format compatibility with OpenSSL was an objective then it will be lost with your new prefix....but maybe that was never intended anyway.

**Reply by lacostej** [reply]
It's possible that `Salted__` was used for compatibility, or just because the author found this as example. As I understand, `openssl enc` [doesn't support AEAD ciphers](https://security.stackexchange.com/questions/266130/generate-aes-256-gcm-key). So if we want to keep that compatibility, I would have to use another cipher.

**Reply by mattcaswell** [reply]
Correct `openssl enc` does not support AEAD ciphers.

**Comment by lacostej** [comment]
Thanks a lot @t8m/@mattcaswell for the detailed explanation. I didn't realize that `bad decrypt` in our context was only reporting the invalid padding and it wasn't based on proper integrity checks.

We understand this part of fastlane's encryption mechanism needs to be overhauled. Not only did it use the very small iteration count in production, it was based on this **deprecated** [ruby API that uses the `EVP_BytesToKey()` method under the hood](https://docs.ruby-lang.org/en/3.3/OpenSSL/Cipher.html#method-i-pkcs5_keyivgen).

We'll change the implementation to a mechanism that uses proper integrity checking and better security overall. Thanks again. Really appreciated.
